packages/block-editor/src/store/selectors.js

```diff
@@ -2385,6 +2385,21 @@ const getAllowedPatternsDependants = ( select ) => ( state, rootClientId ) => [
 	...getInsertBlockTypeDependants( state, rootClientId ),
 ];
 
+const patternsWithParsedBlocks = new WeakMap();
```

Comment:
Maybe it's just me, but this type of manual optimization feels fragile to me. Why do we want to use a `WeakMap` to cache individual patterns rather than relying on composing cached selectors? Based on the implementation in `getAllPatterns`, we create a new array (and new items in the array) every time `getAllPatternsDependants` changes, breaking the memoization here. That is, every time the patterns list changes (new patterns, removed patterns, etc), we have to recompute all these anyway. The `WeakMap` doesn't really have a difference from a list-level selector. I have not tested this though so I could be very wrong here 😅. It just seems like we're aggressively doing memoization with different techniques without an idiomatic mental model (or that I'm lacking it 🤦). Memoization isn't free after all. I wonder if anyone else is feeling lost too 🤔.

Reply:
> I wonder if anyone else is feeling lost sometimes too

Yes, I personally also feel lost all the time. Memoization is indeed extremely hard to understand 100%. It's probably the biggest downside of the Redux architecture and React in general. They depend a lot on exact identity and equality of objects, and the underlying JS language primitives don't have a good support for that.

> Based on the implementation in `getAllPatterns`, we create a new array (and new items in the array) every time `getAllPatternsDependants` changes, breaking the memoization here.

There is a difference between memoizing the entire array, and memoizing the individual items. Getting allowed patterns for different block IDs legitimately produces different arrays, but our issue was that even though the individual patterns are exactly the same, the pattern objects were not the same. Because each call of the selector creates a new object for each array items: adding the `get blocks` enhancement.

> Why do we want to use a `WeakMap` to cache individual patterns rather than relying on composing cached selectors?

A cached selector must have `state` as the first argument, it needs to select from the Redux state. But here we're working only with the `pattern` object, we don't have the state.